midend/flattenUnions.cpp

```diff
@@ -0,0 +1,359 @@
+            }
+        }
+        return new IR::Type_Struct(s->name, s->annotations, fields);
+    }
+    return s;
+}
+
+const IR::Node* DoFlattenHeaderUnion::postorder(IR::Declaration_Variable *dv) {
+    auto ftype = typeMap->getTypeType(dv->type, true);
+    if (ftype->is<IR::Type_HeaderUnion>()) {
+        std::map <cstring, cstring> fieldMap;
+        IR::IndexedVector<IR::Declaration> toInsert;
+        for (auto sfu : ftype->to<IR::Type_HeaderUnion>()->fields) {
+            cstring uName = refMap->newName(dv->name.name + "_" + sfu->name.name);
+            auto uType =  sfu->type->getP4Type();
+            fieldMap.emplace(sfu->name.name, uName);
+            toInsert.push_back(new IR::Declaration_Variable(IR::ID(uName), uType));
+        }
+        replacementMap.emplace(dv->name.name, fieldMap);
+        replaceDVMap.emplace(dv, toInsert);
+    }
+    return dv;
+}
+
+const IR::Node* DoFlattenHeaderUnion::postorder(IR::Member* m) {
+    if (m->expr->type->to<IR::Type_HeaderUnion>()) {
+        if (auto huf = m->expr->to<IR::Member>()) {
+            if (replacementMap.count(huf->member.name)) {
+                if (replacementMap.at(huf->member.name).count(m->member.name)) {
+                    auto newHuName = replacementMap.at(huf->member.name).at(m->member.name);
+                    return new IR::Member(huf->expr,IR::ID(newHuName));
+                }
+            }
+            return m;
+        } else if (auto huf = m->expr->to<IR::PathExpression>()) {
+            if (replacementMap.count(huf->path->name.name)) {
+                if (replacementMap.at(huf->path->name.name).count(m->member.name)) {
+                    auto newHuName = replacementMap.at(huf->path->name.name).at(m->member.name);
+                    return new IR::PathExpression(IR::ID(newHuName));
+                }
+            }
+            return m;
+        }
+    }
+    return m;
+}
+const IR::Node* DoFlattenHeaderUnion::postorder(IR::P4Action* action) {
+    auto actiondecls = action->body->components;
+    for (auto rdv : replaceDVMap) {
+        if (auto decl = action->getDeclByName(rdv.first->name.name)) {
+            for (auto it = actiondecls.begin();it != actiondecls.end(); it++) {
+                auto d = *it;
+                if (d->is<IR::Declaration>() && decl->is<IR::Declaration>()) {
+                    auto dName = d->to<IR::Declaration>()->name.name;
+                    auto declName = decl->to<IR::Declaration>()->name.name;
+                    if (dName == declName) {
+                        actiondecls.insert(it, rdv.second.begin(), rdv.second.end());
+                    }
+                }
+            }
+        }
+    }
+    auto body = new IR::BlockStatement(action->body->srcInfo);
+    for (auto a : actiondecls)
+        body->push_back(a);
+    action->body = body;
+    return action;
+}
+
+const IR::Node* DoFlattenHeaderUnion::postorder(IR::Function* function) {
+    auto functiondecls = function->body->components;
```

Comment:
aren't the functions inlined by now?

Reply:
Yes they are. I removed the visitor and also added a test for this.